GitCommands/UserRepositoryHistory/LocalRepositoryManager.cs

```diff
@@ -314,5 +316,29 @@ private static IEnumerable<Repository> AdjustHistorySize(IEnumerable<Repository>
         {
             return repositories.Take(recentRepositoriesHistorySize);
         }
+
+        public async Task RemoveInvalidRepositoriesAsync(Func<string, bool> predicate)
+        {
+            await TaskScheduler.Default;
+            var recentRepositoryHistory = await LoadRecentHistoryAsync();
+            var invalidRecentRepositories = recentRepositoryHistory
+                                            .Select(repo => repo.Path)
+                                            .Where(path => !predicate(path));
+
+            foreach (var path in invalidRecentRepositories)
+            {
+                await RemoveRecentAsync(path);
```

Comment:
This is very inefficient, since internally `RemoveRecentAsync` performs load and save.
You need to remove all invalid repos from `recentRepositoryHistory` and then save.

Something like this:

```
            await TaskScheduler.Default;
            var repositoryHistory = await LoadRecentHistoryAsync();

            var existingCount = repositoryHistory.Count;

            // remove invalid

            if (existingCount != repositoryHistory.Count)
            {
                await SaveRecentHistoryAsync(repositoryHistory);
            }
```

Reply:
Done.